hazelcast/src/main/java/com/hazelcast/map/impl/MapServiceContextImpl.java

```diff
@@ -436,7 +437,13 @@ public void destroyMap(String mapName) {
         }
 
         nodeEngine.getWanReplicationService().removeWanEventCounters(MapService.SERVICE_NAME, mapName);
-        mapContainer.getMapStoreContext().stop();
+        MapStoreContext mapStoreContext = mapContainer.getMapStoreContext();
+        mapStoreContext.stop();
+        MapStoreWrapper mapStoreWrapper = mapStoreContext.getMapStoreWrapper();
+        //If the map has a MapStore, destroy that as well
+        if (mapStoreWrapper != null) {
+            mapStoreWrapper.destroy();
```

Comment:
Just wanted to note that this change can break existing programs. 

Calling `destroy` here is  against the `MapLoaderLifecycleSupport#destroy` contract which declares destroy will only be called on graceful shutdown: https://github.com/hazelcast/hazelcast/blob/0a88304329138340e91a1affa0f8ada9fc43ede0/hazelcast/src/main/java/com/hazelcast/map/MapLoaderLifecycleSupport.java#L58-L64

Reply:
I think it does not break the contract because destroy() will still be called upon shutdown.

Reply:
It is breaking the contract by not only calling it on shutdown. Now it is also executed on map destroy. Scope is widened. Note that IMap can be destroyed on a running node, node shutdown is not required, this is the part broken.

Maybe as an alternative you can add a new method or interface to lifecycle-support and you can call it on node shutdown.

Reply:
@orcunc we should add a test with the above scenario to probably `MapStoreTest` class, which seems to have the necessary infrastructure.

Reply:
I think it is ok to call `MapLoaderLifecycleSupport#destroy` upon `IMap` destruction. Reasoning:
- `MapLoaderLifecycleSupport` class javadoc discusses initialization and destruction (instead of Hazelcast startup/shutdown):
```
 * MapLoader classes implementing MapLoaderLifecycleSupport
 * interface will be called by Hazelcast on init and destroy so that
 * implementation can do necessary configuration when initializing and
 * clean up when destroying.
```
To me, the intent seems to be to integrate with map lifecycle, not the `HazelcastInstance` lifecycle (tbh I always thought it was that way, until I read this review thread :) ).
- I agree init/destroy is broken as demonstrated by Franta's example above.

Those who implement `MapLoaderLifecycleSupport` can retain the reference to `HazelcastInstance` from `init` and check `instance.getLifecycleService().isRunning()` in `destroy()` impl to react accordingly to `IMap` being destroyed or `HazelcastInstance` shutdown.